File: papers/Structure/Analysis/QG/jl/qg_drifters.jl
"""
    qg_drifters.jl

Advect virtual Lagrangian drifters through QG velocity fields using Drifters.jl.

Depends on: qg_grid.jl (must be included first)

Usage:
    include("qg_grid.jl")
    include("qg_drifters.jl")
    results = run_drifters(; t_start=5001, n_days=30, n_drifters_per_side=64)
"""

using Drifters
using Drifters: DataFrames, DataFrame, groupby, combine, nrow

"""
    load_uv_arrays(nc_path::String, t_idx::Int, dx::Float64, dy::Float64; lev::Int=1)

Load a single velocity snapshot from the NetCDF file, normalize to grid units/s,
and apply C-grid staggering shift.

Returns: (u_grid::Matrix{Float64}, v_grid::Matrix{Float64})
"""
function load_uv_arrays(nc_path::String, t_idx::Int, dx::Float64, dy::Float64; lev::Int=1)
    ds = NCDataset(nc_path, "r")
    u_phys = ds["u"][:, :, lev, t_idx]  # m/s
    v_phys = ds["v"][:, :, lev, t_idx]  # m/s
    close(ds)

    # Normalize to grid units per second
    u_grid = u_phys ./ dx
    v_grid = v_phys ./ dy

    # Apply C-grid shift: pyqg velocities are at cell centers,
    # but Drifters.jl expects C-grid staggering where u[i,j] is at (i-1, j-1/2)
    # and v[i,j] is at (i-1/2, j-1)
    to_C_grid!(u_grid, dims=1)
    to_C_grid!(v_grid, dims=2)

    return u_grid, v_grid
end

"""
    deploy_drifters(n_per_side::Int, nx::Int)

Create initial drifter positions on a regular sub-grid.
Positions are in grid-index coordinates (0 to nx).

Returns: (x0::Vector{Float64}, y0::Vector{Float64})
"""
function deploy_drifters(n_per_side::Int, nx::Int)
    spacing = nx / n_per_side
    offsets = range(spacing/2, stop=nx - spacing/2, length=n_per_side)

    x0 = Float64[]
    y0 = Float64[]
    for yy in offsets, xx in offsets
        push!(x0, xx)
        push!(y0, yy)
    end
    return x0, y0
end

"""
    run_drifters(; nc_path=nothing, t_start=5001, n_days=30,
                   n_per_side=64, lev=1, record_interval=1)

Run drifter advection through the QG flow field.

Arguments:
- nc_path: path to QG NetCDF file (defaults to \$OS_DATA/QG/...)
- t_start: starting time index in the NetCDF (1-based)
- n_days: number of days to advect
- n_per_side: drifters per side of the deployment grid (total = n_per_side^2)
- lev: vertical level (1=upper, 2=lower)
- record_interval: record positions every N days

Returns: NamedTuple with fields:
- trajectories: DataFrame with (ID, x, y, t, x_m, y_m) columns
- dx: grid spacing in meters
- nx: grid size
- n_drifters: total number of drifters
"""
function run_drifters(; nc_path=nothing, t_start::Int=5001, n_days::Int=30,
                       n_per_side::Int=64, lev::Int=1, record_interval::Int=1)
    if isnothing(nc_path)
        nc_path = joinpath(ENV["OS_DATA"], "QG", "QGModelOutput20years.nc")
    end

    # Get grid info
    ds = NCDataset(nc_path, "r")
    x_coord = ds["x"][:]
    nx = length(x_coord)
    dx = x_coord[2] - x_coord[1]
    dy = dx  # square grid
    close(ds)

    dt_seconds = 86400.0  # 1 day between snapshots

    # Deploy drifters
    x0, y0 = deploy_drifters(n_per_side, nx)
    n_drifters = length(x0)
    println("Deploying $n_drifters drifters on $(n_per_side)x$(n_per_side) grid")

    # Initialize trajectory storage
    traj = DataFrame(ID=Int[], x=Float64[], y=Float64[], t=Float64[])

    # Record initial positions
    t_abs = 0.0
    for i in 1:n_drifters
        push!(traj, (ID=i, x=x0[i], y=y0[i], t=t_abs))
    end

    # Load first velocity field
    println("Loading initial velocity field (t_idx=$t_start)...")
    u0, v0 = load_uv_arrays(nc_path, t_start, dx, dy; lev=lev)

    # Time-stepping loop
    for day in 1:n_days
        t_idx_next = t_start + day

        # Load next velocity field
        u1, v1 = load_uv_arrays(nc_path, t_idx_next, dx, dy; lev=lev)

        # Build FlowFields (time spans one day in seconds)
        F = FlowFields(u0, u1, v0, v1, [0.0, dt_seconds])

        # Build Individuals
        I = Individuals(F, x0, y0)

        # Integrate one day
        ∫!(I)

        # Update positions (with periodic wrapping)
        # 📌 is Matrix{Vector{Float64}} of size (1, n_drifters)
        # Each I.📌[1,i] is a [x, y] vector
        for i in 1:n_drifters
            pos = I.📌[1, i]
            x0[i] = mod(pos[1], nx)
            y0[i] = mod(pos[2], nx)
        end

        # Record positions
        t_abs += dt_seconds
        if day % record_interval == 0
            for i in 1:n_drifters
                push!(traj, (ID=i, x=x0[i], y=y0[i], t=t_abs))
            end
        end

        # Advance velocity fields
        u0, v0 = u1, v1

        if day % 10 == 0
            println("  Day $day/$n_days complete")
        end
    end

    # Add physical coordinates
    traj.x_m = traj.x .* dx
    traj.y_m = traj.y .* dy

    println("Done. $(nrow(traj)) trajectory records for $n_drifters drifters over $n_days days.")

    return (trajectories=traj, dx=dx, nx=nx, n_drifters=n_drifters)
end

File: papers/Structure/Analysis/QG/py/qg_drifters.py
"""
Python wrapper for the Julia QG drifter simulation.

Calls the Julia CLI script via subprocess and returns results as pandas DataFrames.

Usage:
    from py.qg_drifters import run_drifters, load_trajectories

    traj = run_drifters(t_start=5001, n_days=30, n_per_side=16)
    print(traj.head())
"""

import subprocess
import shutil
import json
import os
from pathlib import Path

import pandas as pd

# Paths
_THIS_DIR = Path(__file__).parent
_JL_DIR = _THIS_DIR.parent / "jl"
_CLI_SCRIPT = _JL_DIR / "run_drifters_cli.jl"
_DEFAULT_OUTPUT_DIR = Path("/tmp/qg_drifters")


def _find_julia():
    """Find the Julia executable."""
    # Check PATH
    julia = shutil.which("julia")
    if julia:
        return julia
    # Check juliaup default location
    juliaup_julia = Path.home() / ".juliaup" / "bin" / "julia"
    if juliaup_julia.exists():
        return str(juliaup_julia)
    raise FileNotFoundError(
        "Julia not found. Install via: curl -fsSL https://install.julialang.org | sh"
    )


def run_drifters(
    t_start=5001,
    n_days=30,
    n_per_side=16,
    lev=1,
    record_interval=1,
    box_center_x=None,
    box_center_y=None,
    box_size_km=None,
    drifter_spacing_km=None,
    output_path=None,
    cache=True,
    verbose=True,
):
    """Run the Julia drifter simulation and return trajectories as a DataFrame.

    Parameters
    ----------
    t_start : int
        Starting time index in the NetCDF file (1-based). Default 5001 (day ~5000).
    n_days : int
        Number of days to advect.
    n_per_side : int
        Drifters per side of the deployment grid (total = n_per_side^2).
        Ignored when box_size_km is specified.
    lev : int
        Vertical level (1=upper, 2=lower).
    record_interval : int
        Record positions every N days.
    box_center_x : float, optional
        Center x of deployment box in grid units. Default: domain center.
    box_center_y : float, optional
        Center y of deployment box in grid units. Default: domain center.
    box_size_km : float, optional
        Side length of deployment box in km. If None, uses full-domain deployment.
    drifter_spacing_km : float, optional
        Spacing between drifters in km. Default: 10 km (only used with box_size_km).
    output_path : str or Path, optional
        Path for the output CSV. Defaults to /tmp/qg_drifters/<params>.csv.
    cache : bool
        If True, skip re-running if output CSV already exists.
    verbose : bool
        If True, stream Julia stdout in real-time.

    Returns
    -------
    traj : pd.DataFrame
        Trajectory data with columns: ID, x, y, t, x_m, y_m
    meta : dict
        Metadata: dx, nx, n_drifters, t_start, n_days, n_per_side, lev
    """
    # Determine output path
    if output_path is None:
        _DEFAULT_OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
        if box_size_km is not None:
            bcx = box_center_x if box_center_x is not None else "c"
            bcy = box_center_y if box_center_y is not None else "c"
            dsp = drifter_spacing_km if drifter_spacing_km is not None else 10
            fname = f"traj_t{t_start}_d{n_days}_box{box_size_km}km_sp{dsp}km_cx{bcx}_cy{bcy}_l{lev}_r{record_interval}.csv"
        else:
            fname = f"traj_t{t_start}_d{n_days}_n{n_per_side}_l{lev}_r{record_interval}.csv"
        output_path = _DEFAULT_OUTPUT_DIR / fname
    output_path = Path(output_path)
    meta_path = Path(str(output_path) + ".meta.json")

    # Check cache
    if cache and output_path.exists() and meta_path.exists():
        if verbose:
            print(f"Using cached results: {output_path}")
        return load_trajectories(output_path)

    # Find Julia
    julia = _find_julia()

    # Build command
    cmd = [
        julia,
        str(_CLI_SCRIPT),
        "--t_start", str(t_start),
        "--n_days", str(n_days),
        "--n_per_side", str(n_per_side),
        "--lev", str(lev),
        "--record_interval", str(record_interval),
        "--output", str(output_path),
    ]
    if box_size_km is not None:
        cmd.extend(["--box_size_km", str(box_size_km)])
    if box_center_x is not None:
        cmd.extend(["--box_center_x", str(box_center_x)])
    if box_center_y is not None:
        cmd.extend(["--box_center_y", str(box_center_y)])
    if drifter_spacing_km is not None:
        cmd.extend(["--drifter_spacing_km", str(drifter_spacing_km)])

    if verbose:
        print(f"Running: {' '.join(cmd[:3])} ...")
        print(f"Output: {output_path}")

    # Run Julia with real-time output streaming
    process = subprocess.Popen(
        cmd,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
        bufsize=1,
    )

    output_lines = []
    for line in process.stdout:
        output_lines.append(line)
        if verbose:
            print(f"  [julia] {line}", end="")

    returncode = process.wait()
    if returncode != 0:
        full_output = "".join(output_lines)
        raise RuntimeError(
            f"Julia drifter simulation failed (exit code {returncode}):\n{full_output}"
        )

    return load_trajectories(output_path)


def load_trajectories(csv_path):
    """Load trajectories and metadata from a previously saved run.

    Parameters
    ----------
    csv_path : str or Path
        Path to the trajectory CSV file.

    Returns
    -------
    traj : pd.DataFrame
        Trajectory data.
    meta : dict
        Metadata dictionary.
    """
    csv_path = Path(csv_path)
    meta_path = Path(str(csv_path) + ".meta.json")

    traj = pd.read_csv(csv_path)

    meta = {}
    if meta_path.exists():
        with open(meta_path) as f:
            meta = json.load(f)

    return traj, meta
